ecsmetadata/client.go

```diff
@@ -0,0 +1,205 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License").
+// You may not use this file except in compliance with the License.
+// A copy of the License is located at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package ecsmetadata queries ECS Metadata Server for ECS task metrics.
+// This package is currently experimental and is subject to change.
+package ecsmetadata
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+)
+
+var endpoint string
+
+func init() {
+	const endpointEnv = "ECS_CONTAINER_METADATA_URI_V4"
+	endpoint = os.Getenv(endpointEnv)
+	if endpoint == "" {
+		log.Fatalf("%q environmental variable is not set, are you running this on ECS?", endpointEnv)
+	}
+}
+
+type Client struct {
+	// httpClient is the client to use when making HTTP requests
+	// when set. Otherwise, an internal default client will be used.
+	httpClient *http.Client
+}
+
+func NewClient(httpClient *http.Client) (*Client, error) {
+	if httpClient == nil {
+		httpClient = &http.Client{}
+	}
+	return &Client{httpClient: httpClient}, nil
+}
+
+func (c *Client) RetrieveTaskStats(ctx context.Context) (map[string]*ContainerStats, error) {
+	out := make(map[string]*ContainerStats)
+	err := c.request(ctx, endpoint+"/task/stats", &out)
+	return out, err
+}
+
+func (c *Client) RetrieveTaskMetadata(ctx context.Context) (*TaskMetadata, error) {
+	var out TaskMetadata
+	err := c.request(ctx, endpoint+"/task", &out)
+	// Replace cluster ARN with cluster name.
+	if index := strings.LastIndex(out.Cluster, "/"); index > 0 {
+		out.Cluster = out.Cluster[index+1:]
+	}
+	// TODO(jbd): Consider using taskWithTags for task tags.
+	return &out, err
+}
+
+func (c *Client) request(ctx context.Context, uri string, out interface{}) error {
+	req, err := http.NewRequest("GET", uri, nil)
+	if err != nil {
+		return err
+	}
+	req = req.WithContext(ctx)
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	log.Printf("%s", body)
```

Comment:
This is probably a debug statement.

Reply:
Put it intentionally to have debug logs but removing it for now. We can decide whether we want to have any debug logs.